### PR TITLE
Add Column Start and Row Start controls to Grid children

### PR DESCRIPTION
**Pull request by noisysocks**
## What?
Largely follow https://github.com/WordPress/gutenberg/pull/58539. Part of https://github.com/WordPress/gutenberg/issues/57478.

Adds Column Start and Row Start controls to the block inspector for items that are a child of grid layout.

Also rejigs the grouping of the existing Tool Panel items so that we have _Grid position_ and _Grid span_ in the menu. See https://github.com/WordPress/gutenberg/pull/59488. Props @tellthemachines.

## Why?
As part of https://github.com/WordPress/gutenberg/issues/57478 I'm working on letting the user drag and drop items within a grid to position them manually. For this to work we need to be able to set `style.layout.columnStart` and `style.layout.rowStart` and we'll need a keyboard accessible alternative for setting these attributes. This PR adds both.

## How?
Most of the work is already done by https://github.com/WordPress/gutenberg/pull/58539. I'm just adjusting how we output `grid-column` and `grid-row` and adding two new `InputControls`.

## Testing Instructions
1. Insert a Grid block.
2. Add some child blocks to the Grid.
3. Select one of the children. 
4. The Column Start and Row Start controls should appear  in the block inspector.

## Screenshots or screencast 

https://github.com/WordPress/gutenberg/assets/612155/a6a15279-c62e-455b-a993-1d5bd8d96407

### CODE REVIEW

File: lib/block-supports/layout.php [hunk 1]
```diff
@@ -581,11 +581,25 @@ function gutenberg_render_layout_support_flag( $block_content, $block ) {
 		$child_layout_declarations['flex-grow'] = '1';
 	}
 
-	if ( isset( $block['attrs']['style']['layout']['columnSpan'] ) ) {
+	if ( isset( $block['attrs']['style']['layout']['columnStart'] ) && isset( $block['attrs']['style']['layout']['columnSpan'] ) ) {
```
Comment (tellthemachines) [inline lib/block-supports/layout.php]: We could reduce the uses of `isset` (and make these lines a bit easier on the eye 😅 ) if we assign these to variables outside of the conditions, e.g.:

`$column_start  = isset( $block['attrs']['style']['layout']['columnStart'] ) ? $block['attrs']['style']['layout']['columnStart']  : null`

**Review by tellthemachines — COMMENTED** [review]
Positioning the items is working really well! One thing that is starting to be bothersome is that we can't reset each property individually:
Screenshot 2024-03-01 at 11 45 08 am
My bad for not making it part of the initial implementation - it didn't seem like such an issue when all we had was Col and Row span, but it's definitely feeling off now. If you don't feel like adding a bunch of extra logic to this PR I'm happy to address it as a follow-up!

**Comment by noisysocks** [comment]
Yeah true @tellthemachines it would be good to split them and I think have "Grid starts" hidden by default.

File: lib/block-supports/layout.php (hunk 1 above)
Comment (noisysocks) [reply lib/block-supports/layout.php]: But if they're not set then won't that result in a PHP `undefined index` notice?

Comment (noisysocks) [reply lib/block-supports/layout.php]: Oh you covered that with the ternary. Cool 👍 I'll do that

**Comment by tellthemachines** [comment]
Just pushed a fix for an issue with columnStart in Auto mode, where once the column number goes below the start number, the grid will push the item into an auto column with no minimum size, and won't reflow gracefully:

Screenshot 2024-03-01 at 4 07 54 pm

I adjusted the existing container query so that, if a columnStart is present, once the number of columns goes below it the item will either span full-width if it also has a columnSpan or get reset if it doesn't. I _think_ that's the best way to deal with it but open to other suggestions!

For Manual grids, should we be checking whether the columnStart is less than the columnCount? There can be unexpected effects if it goes over 😅 

Had to rebase to avoid conflicts with https://github.com/WordPress/gutenberg/commit/0b5cd48bfb474377e6947a7c3302d37bd910a586.

**Comment by noisysocks** [comment]
> For Manual grids, should we be checking whether the columnStart is less than the columnCount? There can be unexpected effects if it goes over 😅

Sounds good. Should we set `max` on the input? Or do we clamp the value before rendering it in CSS?

**Comment by tellthemachines** [comment]
> Should we set `max` on the input? Or do we clamp the value before rendering it in CSS?

I'm inlined towards setting on the input so it's clear to the users, instead of letting them set a value and then changing it.

**Comment by noisysocks** [comment]
Oops forgot to `git push` yesterday. I've added a `max` to the inputs set to the number of columns when the layout is in manual mode.

**Comment by tellthemachines** [comment]
I'm seeing the controls go all funny when grid is in Manual mode, can you repro this?
Screenshot 2024-03-05 at 11 19 29 am

**Comment by noisysocks** [comment]
> I adjusted the existing container query so that, if a columnStart is present, once the number of columns goes below it the item will either span full-width if it also has a columnSpan or get reset if it doesn't. I _think_ that's the best way to deal with it but open to other suggestions!

I tested this and it works as you describe.

I'm not sure about the behaviour though. I feel that in this case we can't possibly know what the user actually wants to do and so anything we do is a guess and will likely be unexpected. 

Is it useful to set a start when the grid is in Auto mode? Maybe we make it so that column start / row start can only be set when the parent layout is a Manual grid? And then clamp the start values so that `1 ≤ start column/row ≤ num columns/rows`?

> I'm seeing the controls go all funny when grid is in Manual mode, can you repro this?

I'm not seeing this 😅

**Comment by tellthemachines** [comment]
> I'm not sure about the behaviour though. I feel that in this case we can't possibly know what the user actually wants to do and so anything we do is a guess and will likely be unexpected.

I think this responsive behaviour is less unexpected than what happens if we don't do anything 😅 
When an item is set to a column-start that's bigger than the actual number of columns in the grid, the grid creates some auto-columns to fit it in the place it wants to be. But the auto-columns have no particular width by default, so they end up shrinking to the minimum possible size. 

So let's say we have an auto grid with the black-and-white cat pic pinned to the 4th col:
Screenshot 2024-03-05 at 1 28 27 pm

With no responsiveness, if there's not enough space for a 4th col, the grid adds it all the same but makes it tiny:
Screenshot 2024-03-05 at 1 29 15 pm

Whereas with the container query, it looks a lot more sensible even if we did have to unpin the item:
Screenshot 2024-03-05 at 1 30 10 pm

I'm not super concerned with the unexpectedness because choosing an auto grid setting implies some level of giving up control on the exact position of things already 😅 so this is really just to ensure that no bad breakage occurs.

Setting the grid spans to `1 / -1` once the grid is too small for `span` number of cols assumes that if you're making an item span multiple columns you'll still want it to be bigger than the other items on smaller screens. That's a bit more of an assumption, but ultimately we do have to choose some default here and in testing that one seemed to work more gracefully than just plain resetting to `auto`.

> Is it useful to set a start when the grid is in Auto mode? Maybe we make it so that column start / row start can only be set when the parent layout is a Manual grid? And then clamp the start values so that `1 ≤ start column/row ≤ num columns/rows`?

I think it is still useful, for pinning items to the top for example.

**Comment by noisysocks** [comment]
I merged https://github.com/WordPress/gutenberg/pull/59488 into this. Easier I think to test both changes together as part of a cohesive update to the child layout controls.

**Comment by noisysocks** [comment]
I also:

- Updated the copy to _Grid placement_ instead of _Grid starts_.
- Made _Grid placement_ off by default as it's more of an advanced use case.

File: packages/block-editor/src/hooks/layout-child.js [hunk 2]
```diff
@@ -44,9 +59,20 @@ function useBlockPropsChildLayoutStyles( { style } ) {
...
-			// Calculate the container query value.
-			const columnSpanNumber = parseInt( columnSpan );
+		if (
+			( columnSpan || columnStart ) &&
+			( minimumColumnWidth || ! columnCount )
+		) {
+			// Check if columnSpan and columnStart are numbers so Math.max doesn't break.
+			const columnSpanNumber = columnSpan ? parseInt( columnSpan ) : null;
+			const columnStartNumber = columnStart
+				? parseInt( columnStart )
```
Comment (andrewserong) [inline packages/block-editor/src/hooks/layout-child.js:69]: Just a drive-by comment, do we need to check that `parseInt` was successful (i.e. an `isNaN` check or something similar to how `parentColumnValue` is checked), or can we always trust that if it was truthy, it'll be a number? Same with the line below.

Comment (tellthemachines) [reply packages/block-editor/src/hooks/layout-child.js:69]: If parseInt isn't successful it returns NaN, which will evaluate to false. Any other result should be a number 😅

Comment (andrewserong) [reply packages/block-editor/src/hooks/layout-child.js:69]: Ah, gotcha. I noticed the value winds up being used directly in the calculation for `containerQueryValue` so wondered if there was a chance `NaN` could be unexpectedly output in the container query line? Not sure if that would ever be reached in practice, though, so feel free to ignore this comment for now!

**Comment by noisysocks** [comment]
@tellthemachines and I were chatting and noted that one issue with letting one set _Grid placement_ is that it doesn't update the DOM order which is not ideal for accessibility. We might need to update the block order when setting placement and set/reset the placement when re-ordering blocks using the movers.

Instead of trying to anticipate what the perfect solution is I've opted to put the UI for setting _Grid placement_ behind the _Grid interactivity_ experimental flag so that we can move forward with https://github.com/WordPress/gutenberg/pull/59490 and continue to iterate.

**Review by andrewserong — COMMENTED** [review]
This is generally testing well for me, but I noticed a couple of UI issues that would be good to tweak.

* In the ToolsPanel menu, `Padding` appears between the two grid items in the menu. It'd be better for the two grid items to be next to each other.

image

* The styling of the input fields appears to be off as they're not filling the available area:

image

Otherwise testing well, and the UI appears to be safely guarded behind the experiment flag. One potential idea could be to also display the UI in the inspector controls if a value is set but the experiment is switched off, so that if folks who aren't using the experiment encounter markup that has values set, it's possible to clear out values? That might not be necessary, though, so just an idea 🙂

**Comment by tellthemachines** [comment]
> The styling of the input fields appears to be off as they're not filling the available area

Oh, that's the same issue I was seeing last week! Only on the manual grid, right? But @noisysocks couldn't repro it.

**Comment by andrewserong** [comment]
> Only on the manual grid, right?

Yep, only on manual.

File: packages/block-editor/src/components/child-layout-control/index.js [hunk 3]
```diff
@@ -104,37 +151,93 @@ export default function ChildLayoutControl( {
...
+							as={ ToolsPanelItem }
+							hasValue={ hasStartValue }
+							label={ __( 'Grid placement' ) }
+							onDeselect={ resetGridStarts }
+							isShownByDefault={ false }
+							panelId={ panelId }
+						>
+							<InputControl
+								size={ '__unstable-large' }
+								label={ __( 'Column' ) }
```
Comment (andrewserong) [inline packages/block-editor/src/components/child-layout-control/index.js]: Oh, just double-checking, should the visual label on these control also use the word `placement`, so that it matches the menu item?

Comment (tellthemachines) [reply packages/block-editor/src/components/child-layout-control/index.js]: It would make things clearer! It's a pretty long string, though I'm not sure how relevant that is given that strings can have any length in translation.

**Comment by noisysocks** [comment]
> > The styling of the input fields appears to be off as they're not filling the available area
> 
> Oh, that's the same issue I was seeing last week! Only on the manual grid, right? But @noisysocks couldn't repro it.

Turns out this is webpack specific. It seems related to the `max` attribute on the `<input>` affecting the intrinsic width of the input.

https://stackoverflow.com/questions/33283901/input-number-max-attribute-resizes-field

I'm not sure how to fix this at the `@wordpress/components` level (adding `width: fill-available` as suggested above does not work) so instead I worked around the issue by giving an explicit `width: 50%` style to the flex items.

File: packages/block-editor/src/components/child-layout-control/index.js (hunk 3 above)
Comment (noisysocks) [reply packages/block-editor/src/components/child-layout-control/index.js]: I don't think it's necessary. Also if we say _Column placement_ and _Row placement_ the labels truncate as they're too long.

I suspect we'll change the labels (and lots of other things) based on user testing.

**Comment by noisysocks** [comment]
> * In the ToolsPanel menu, `Padding` appears between the two grid items in the menu. It'd be better for the two grid items to be next to each other.

I agree but I'm not sure if it's possible? It looks like `ToolsPanel` will output all of the non-optional items and then all of the optional items, and each group is arranged in the order they're registered. We might need to adjust how `ToolsPanel` works but I don't want to do that in this PR.

**Review by tellthemachines — APPROVED** [review]
Given this is an experiment, I think it's in a mergeable state now ✅ 

We will have to address the issue of matching visual and markup order before stabilising.
